domain/src/com/thoughtworks/go/domain/materials/git/GitCommand.java

```diff
@@ -69,14 +72,45 @@ public int clone(ProcessOutputStreamConsumer outputStreamConsumer, String url) {
     // Clone repository from url with specified depth.
     // Special depth 2147483647 (Integer.MAX_VALUE) are treated as full clone
     public int clone(ProcessOutputStreamConsumer outputStreamConsumer, String url, Integer depth) {
+        CommandLine gitClone = getGitCloneCommand(outputStreamConsumer, url, depth);
+        return run(gitClone, outputStreamConsumer);
+    }
+
+    protected CommandLine getGitCloneCommand(ProcessOutputStreamConsumer outputStreamConsumer, String url, Integer depth) {
         CommandLine gitClone = cloneCommand();
 
         if(depth < Integer.MAX_VALUE) {
             gitClone.withArg(String.format("--depth=%s", depth));
+            if(isSubModuleShallowCloneSupported()){
+                outputStreamConsumer.stdOutput("[GIT] Recursively shallow cloning submodule repositories");
+                gitClone.withArg("--recursive").withArg("--recurse-submodules").withArg("--shallow-submodules");
+            }
         }
         gitClone.withArg(new UrlArgument(url)).withArg(workingDir.getAbsolutePath());
+        return gitClone;
+    }
 
-        return run(gitClone, outputStreamConsumer);
+    protected boolean isSubModuleShallowCloneSupported() {
+        String version = GitCommand.version(new HashMap<String, String>());
+        //--shallow-submodules flag is supported from git version 2.9.0
+        return isVersionEqualToOrHigherThan(version, 2.9f);
+    }
+
+    public static boolean isVersionEqualToOrHigherThan(String hgout, float version) {
```

Comment:
Change the argument name to something that indicates that its the output of the command git version.